File: 2023/day12/task2.py

import functools
from typing import Tuple

@functools.cache
def find_permutations(remaining_input : str, remaining_pattern:Tuple[int, ...], current_defects):
    # print(remaining_input,remaining_pattern,current_defects)
    if len(remaining_input)==0:
        if len(remaining_pattern) == 0:
            return 1
        elif len(remaining_pattern) == 1:
            return 1 if current_defects==remaining_pattern[0] else 0
        else:
            return 0
    if len(remaining_pattern) == 0:
        return 0 if '#' in remaining_input else 1

    first = remaining_input[0]
    current_pattern = remaining_pattern[0]
    if first == '?':
        if current_defects==0:
            return find_permutations(remaining_input[1:],remaining_pattern,0) + find_permutations(remaining_input[1:],remaining_pattern,1)
        if current_defects == current_pattern:
            return find_permutations(remaining_input[1:],remaining_pattern[1:],0)
        return find_permutations(remaining_input[1:],remaining_pattern,current_defects+1)
    
    if first == '.':
        if current_defects==0:
            return find_permutations(remaining_input[1:],remaining_pattern,0)
        if current_defects == current_pattern:
            return find_permutations(remaining_input[1:],remaining_pattern[1:],0)
        else:
            return 0
        
    if first == '#':
        return find_permutations(remaining_input[1:],remaining_pattern,current_defects+1)

# ???.### 1,1,3
def prepare_line(line:str):
    line = line.strip()
    input, pattern = line.split(" ")
    pattern = [int(x) for x in pattern.split(',')]
    #return input,tuple(pattern) # Task 1
    return ((input+'?')*5)[:-1],tuple(pattern*5)


def solve():
    with open("input.txt") as input:
        lines = input.readlines()
        total = 0
        for line in lines:
            input,pattern = prepare_line(line)
            perms = find_permutations(input,pattern,0)
            # print(line,"=>",perms)
            total+= perms
        print(total)

solve()

File: 2023/day12/src/bin/task2.rs


fn number_of_valid_permutations(line : &[char], pattern: &[usize]) -> usize {
    
    if pattern.len() == 0 {
        if !line.contains(&'#') {
            return 1;
        }
        return 0;
    }
    
    if line.len() == 0 {
        return 0;
    }

    let first_char = *line.first().unwrap();
    if  first_char == '.'{
        return number_of_valid_permutations(&line[1..], pattern);
    }
    else if first_char == '#' {

        let number_of_defects = *pattern.first().expect("Should have first element");
        if line.starts_with(&['#'].repeat(number_of_defects)) {
            return number_of_valid_permutations(&line[number_of_defects..],&pattern[1..]);
        }
        return 0;
    }
    else if first_char == '?' {
        let mut permuations = 0;
    }
    return 0;
}


// ?###???????? 3,2,1
fn solve_line(line : &str) -> usize {
    let split = line.split(" ").collect::<Vec<&str>>();
    let springs = split[0].to_string();
    let pattern_string = split[1];
    let pattern = pattern_string.split(",").map(|s| s.parse::<usize>().expect("Should be number")).collect::<Vec<usize>>();
    let springs = springs.chars().collect::<Vec<char>>();
    return number_of_valid_permutations(&springs, &pattern);
    
}


fn solve(input :&str) -> usize{
    let result = input.lines().map(solve_line).sum::<usize>();
    return result;
}

fn main() {

    let input = include_str!("../../input.txt");
    let result = solve(input);
    println!("Result1: {}",result);
    let input = include_str!("../../input2.txt");
    let result = solve(input);
    println!("Result2: {}",result);
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn line321() {
        let example = "?###???????? 3,2,1";
        assert_eq!(solve_line(example), 10);
    }
    #[test]
    fn example() {
        let example = include_str!("../../example.txt");
        assert_eq!(solve(example), 21);
    }

    #[test]
    fn patterns() {
        let example = ".###.##.#...";
        let pattern = vec![3,2,1];
        println!("Example 2:");
        let ex2 = vec!['#', '.', '#', '.', '#', '#', '#'];
        let pattern = vec![1, 1, 3];
    }
    
}